File: app/src/main/java/com/teambrella/android/ui/claim/ClaimFragment.java
package com.teambrella.android.ui.claim;

import android.content.res.Resources;
import android.net.Uri;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.app.FragmentManager;
import android.support.v4.app.FragmentTransaction;
import android.text.format.DateUtils;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;
import android.widget.Toast;

import com.google.gson.JsonObject;
import com.teambrella.android.R;
import com.teambrella.android.api.TeambrellaModel;
import com.teambrella.android.api.model.json.JsonWrapper;
import com.teambrella.android.api.server.TeambrellaServer;
import com.teambrella.android.api.server.TeambrellaUris;
import com.teambrella.android.image.TeambrellaImageLoader;
import com.teambrella.android.ui.base.ADataFragment;
import com.teambrella.android.ui.base.ADataProgressFragment;
import com.teambrella.android.ui.chat.ChatActivity;
import com.teambrella.android.ui.widget.ImagePager;

import java.util.ArrayList;

import io.reactivex.Notification;
import jp.wasabeef.picasso.transformations.MaskTransformation;

/**
 * Claim fragment
 */
public class ClaimFragment extends ADataProgressFragment<IClaimActivity> {

    private static final String DETAILS_FRAGMENT_TAG = "details";
    private static final String VOTING_FRAGMENT_TAG = "voting";

    private ImagePager mClaimPictures;
    private ImageView mOriginalObjectPicture;
    private TextView mMessageTitle;
    private TextView mMessageText;
    private TextView mUnreadCount;
    private TextView mWhen;
    private View mDiscussion;

    @Override
    protected View onCreateContentView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_claim, container, false);
        mClaimPictures = view.findViewById(R.id.image_pager);
        mOriginalObjectPicture = view.findViewById(R.id.object_picture);
        mMessageTitle = view.findViewById(R.id.message_title);
        mMessageText = view.findViewById(R.id.message_text);
        mUnreadCount = view.findViewById(R.id.unread);
        mDiscussion = view.findViewById(R.id.discussion);
        mWhen = view.findViewById(R.id.when);


        view.findViewById(R.id.swipe_to_refresh).setEnabled(false);

        if (savedInstanceState == null) {
            mDataHost.load(mTags[0]);
            setContentShown(false);
        }
        return view;
    }

    /**
     * Called immediately after {@link #onCreateView(LayoutInflater, ViewGroup, Bundle)}
     * has returned, but before any saved state has been restored in to the view.
     * This gives subclasses a chance to initialize themselves once
     * they know their view hierarchy has been completely created.  The fragment's
     * view hierarchy is not however attached to its parent at this point.
     *
     * @param view               The View returned by {@link #onCreateView(LayoutInflater, ViewGroup, Bundle)}.
     * @param savedInstanceState If non-null, this fragment is being re-constructed
     */
    @Override
    public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        FragmentManager fragmentManager = getChildFragmentManager();
        FragmentTransaction transaction = fragmentManager.beginTransaction();
        if (fragmentManager.findFragmentByTag(DETAILS_FRAGMENT_TAG) == null) {
            transaction.add(R.id.details_container, ClaimDetailsFragment.getInstance(mTags), DETAILS_FRAGMENT_TAG);
        }

        if (fragmentManager.findFragmentByTag(VOTING_FRAGMENT_TAG) == null) {
            transaction.add(R.id.voting_container, ADataFragment.getInstance(mTags, ClaimVotingFragment.class), VOTING_FRAGMENT_TAG);
        }


        if (!transaction.isEmpty()) {
            transaction.commit();
        }


    }

    @Override
    protected void onDataUpdated(Notification<JsonObject> notification) {
        if (notification.isOnNext()) {
            JsonWrapper response = new JsonWrapper(notification.getValue());
            JsonWrapper data = response.getObject(TeambrellaModel.ATTR_DATA);
            JsonWrapper claimBasic = data.getObject(TeambrellaModel.ATTR_DATA_ONE_BASIC);

            final String smallPhoto;

            if (claimBasic != null) {
                ArrayList<String> photos = TeambrellaModel.getImages(TeambrellaServer.BASE_URL,
                        claimBasic.getObject(), TeambrellaModel.ATTR_DATA_BIG_PHOTOS);
                if (photos != null && photos.size() > 0) {
                    mClaimPictures.init(getChildFragmentManager(), photos);
                    smallPhoto = photos.get(0);
                } else {
                    smallPhoto = null;
                }
                getActivity().setTitle(claimBasic.getString(TeambrellaModel.ATTR_DATA_MODEL));
            } else {
                smallPhoto = null;
            }

            JsonWrapper claimDiscussion = data.getObject(TeambrellaModel.ATTR_DATA_ONE_DISCUSSION);
            if (claimDiscussion != null) {
                mMessageTitle.setText(getString(R.string.claim_title_format_string, data.getInt(TeambrellaModel.ATTR_DATA_ID, 0)));
                String text = claimDiscussion.getString(TeambrellaModel.ATTR_DATA_ORIGINAL_POST_TEXT);

                if (text != null) {
                    mMessageText.setText(text);
                }


                int unread = claimDiscussion.getInt(TeambrellaModel.ATTR_DATA_UNREAD_COUNT, 0);
                mUnreadCount.setText(Integer.toString(unread));
                mUnreadCount.setVisibility(unread > 0 ? View.VISIBLE : View.GONE);

                String objectPhoto = claimDiscussion.getString(TeambrellaModel.ATTR_DATA_SMALL_PHOTO);
                if (objectPhoto != null) {
                    Resources resources = getResources();
                    TeambrellaImageLoader.getInstance(getContext()).getPicasso()
                            .load(TeambrellaServer.BASE_URL + objectPhoto).resize(resources.getDimensionPixelSize(R.dimen.claim_object_picture_with), resources.getDimensionPixelSize(R.dimen.claim_object_picture_height)).centerCrop()
                            .transform(new MaskTransformation(getContext(), R.drawable.teammate_object_mask)).into(mOriginalObjectPicture);
                }


                long now = System.currentTimeMillis();
                long when = now - 60000 * claimDiscussion.getInt(TeambrellaModel.ATTR_DATA_SINCE_LAST_POST_MINUTES);
                mWhen.setText(DateUtils.getRelativeTimeSpanString(when, now, DateUtils.MINUTE_IN_MILLIS, DateUtils.FORMAT_ABBREV_RELATIVE));


                final int claimId = data.getInt(TeambrellaModel.ATTR_DATA_ID);
                final String topicId = claimDiscussion.getString(TeambrellaModel.ATTR_DATA_TOPIC_ID);
                final Uri uri = claimId > 0 ? TeambrellaUris.getClaimChatUri(claimId) : null;


                if (uri != null) {
                    mDiscussion.setOnClickListener(v -> ChatActivity.startClaimChat(getContext()
                            , 0
                            , claimId
                            , claimBasic != null ? claimBasic.getString(TeambrellaModel.ATTR_DATA_MODEL) : null
                            , smallPhoto != null ? Uri.parse(smallPhoto) : null
                            , topicId));
                }
            }

            JsonWrapper claimVoting = data.getObject(TeambrellaModel.ATTR_DATA_ONE_VOTING);
            if (claimVoting != null) {
                View view = getView();
                if (view != null) {
                    view.findViewById(R.id.voting_container).setVisibility(View.VISIBLE);
                }
            }
        } else {
            Toast.makeText(getContext(), "Error", Toast.LENGTH_SHORT).show();
        }
        setContentShown(true);
    }
}

File: app/src/main/java/com/teambrella/android/ui/claim/ClaimFragment.kt
package com.teambrella.android.ui.claim

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import com.google.gson.JsonObject
import com.teambrella.android.R
import com.teambrella.android.api.*
import com.teambrella.android.api.server.TeambrellaUris
import com.teambrella.android.ui.base.ADataProgressFragment
import com.teambrella.android.ui.base.createDataFragment
import com.teambrella.android.ui.chat.ChatActivity
import com.teambrella.android.ui.util.setImage
import com.teambrella.android.ui.widget.ImagePager
import com.teambrella.android.util.ConnectivityUtils
import com.teambrella.android.util.TeambrellaDateUtils
import io.reactivex.Notification
import java.util.*

/**
 * Claim fragment
 */
class ClaimFragment : ADataProgressFragment<IClaimActivity>() {

    companion object {
        private const val DETAILS_FRAGMENT_TAG = "details"
        private const val PAYOUT_FRAGMENT_TAG = "payout"
        private const val VOTING_FRAGMENT_TAG = "voting"
    }

    private val claimPicturesView: ImagePager? by ViewHolder(R.id.image_pager)
    private val originalObjectPictureView: ImageView? by ViewHolder(R.id.object_picture)
    private val messageTitleView: TextView? by ViewHolder(R.id.message_title)
    private val messageTextView: TextView? by ViewHolder(R.id.message_text)
    private val unreadCountView: TextView? by ViewHolder(R.id.unread)
    private val whenView: TextView? by ViewHolder(R.id.`when`)
    private val discussionForeground: View? by ViewHolder(R.id.discussion_foreground)
    private val detailsContainer: View? by ViewHolder(R.id.details_container)
    private val payoutContainer: View? by ViewHolder(R.id.payout_container)

    private var isShown: Boolean = false
    private var teamAccessLevel = TeambrellaModel.TeamAccessLevel.FULL_ACCESS

    override fun onCreateContentView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View {
        val view = inflater.inflate(R.layout.fragment_claim, container, false)
        view.findViewById<View>(R.id.swipe_to_refresh).isEnabled = false
        dataHost.load(tags[0])
        setContentShown(false)
        return view
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        childFragmentManager.apply {
            val transaction = beginTransaction()
            if (findFragmentByTag(DETAILS_FRAGMENT_TAG) == null) {
                transaction.add(R.id.details_container
                        , createDataFragment(tags, ClaimDetailsFragment::class.java)
                        , DETAILS_FRAGMENT_TAG)
            }
    
            if (findFragmentByTag(PAYOUT_FRAGMENT_TAG) == null) {
                transaction.add(R.id.payout_container
                                , createDataFragment(tags, ClaimPayoutFragment::class.java)
                                , PAYOUT_FRAGMENT_TAG)
            }

            if (findFragmentByTag(VOTING_FRAGMENT_TAG) == null) {
                transaction.add(R.id.voting_container
                        , getInstance(tags, MODE_CLAIM)
                        , VOTING_FRAGMENT_TAG)
            }

            if (!transaction.isEmpty) {
                transaction.commit()
            }
        }
        isShown = false
        super.onViewCreated(view, savedInstanceState)
    }

    override fun onDataUpdated(notification: Notification<JsonObject>) {
        if (notification.isOnNext) {
            notification.value?.data?.let {
                val basic = it.basic
                val team = it.team
                val discussion = it.discussionPart
                val photos = basic.bigPhotos?.mapTo(ArrayList()) { it.asString }
                val smallPhoto = if (photos?.isNotEmpty() == true) photos[0] else null

                if (photos != null && photos.isNotEmpty()) {
                    claimPicturesView?.init(childFragmentManager, photos)
                }

                messageTitleView?.text = getString(R.string.claim_title_format_string, it.intId)

                discussion?.originalPostText?.let {
                    messageTextView?.text = it.replace("<p>".toRegex(), "").replace("</p>".toRegex(), "")
                }

                val unreadCount = discussion?.unreadCount ?: 0
                unreadCountView?.text = unreadCount.toString()
                unreadCountView?.visibility = if (unreadCount > 0) View.VISIBLE else View.GONE

                discussion?.smallPhoto?.let {
                    originalObjectPictureView?.setImage(imageLoader.getImageUrl(it), R.dimen.rounded_corners_4dp)
                }

                whenView?.text = TeambrellaDateUtils.getRelativeTime(-(discussion.sinceLastPostMinutes
                        ?: 0L))

                val claimId = it.intId ?: 0
                val uri = if (claimId > 0) TeambrellaUris.getClaimChatUri(claimId) else null
                teamAccessLevel = team?.teamAccessLevel ?: teamAccessLevel

                if (uri != null) {
                    discussionForeground?.setOnClickListener {
                        dataHost.launchActivity(ChatActivity.getClaimChat(context, dataHost.teamId, claimId, basic.model
                                , smallPhoto, discussion?.topicId, teamAccessLevel, basic.incidentDate))
                    }
                }
                
                val showPayout = (basic.votingRes ?: 0.0) > 0.0001
                payoutContainer?.visibility = if (showPayout) View.VISIBLE else View.GONE
                detailsContainer?.setBackgroundResource(if (showPayout) R.drawable.block else R.drawable.block_last)
            }
            setContentShown(true)
            isShown = true
        } else {
            setContentShown(true, !isShown)
            dataHost.showSnackBar(if (ConnectivityUtils.isNetworkAvailable(context)) R.string.something_went_wrong_error else R.string.no_internet_connection)
        }
    }
}

